File: FoosballGames/FoosballGamesQueryHandler.cs
using FoosballGames.Contracts;
using FoosballGames.Infrastructure.Messaging;

namespace FoosballGames;

public class FoosballGamesQueryHandler : IQueryHandler<GetFoosballGames, FoosballGamesResponse>,
    IQueryHandler<GetFoosballGame, Contracts.FoosballGame>
{
    private readonly IFoosballGamesRepository _foosballGamesRepository;

    public FoosballGamesQueryHandler(IFoosballGamesRepository foosballGamesRepository)
    {
        _foosballGamesRepository = foosballGamesRepository;
    }

    public async Task<FoosballGamesResponse> HandleAsync(GetFoosballGames query)
    {
        var results = await _foosballGamesRepository.GetAll();
        var ordered = results.Select(x => x.ToContract()).OrderByDescending(s => s.Start).ToArray();
        return new FoosballGamesResponse(ordered);
    }

    public async Task<Contracts.FoosballGame> HandleAsync(GetFoosballGame query)
    {
        var result = await _foosballGamesRepository.Get(query.Id);
        return result.ToContract();
    }
}

File: FoosballGames/FoosballGamesQueryHandler.fs
namespace FoosballGames

open FoosballGames
open FoosballGames.Contracts
open FoosballGames.Infrastructure.Messaging
open TypeMappers

type FoosballGamesQueryHandler(repository: IFoosballGamesRepository) =
    interface IQueryHandler<GetFoosballGame, FoosballGame> with
        member this.HandleAsync(query) =
            task {
                let! result = repository.Get query.Id
                return result.ToContract()
            }

    interface IQueryHandler<GetFoosballGames, FoosballGamesResponse> with
        member this.HandleAsync _ =
            task {
                let! results = repository.GetAll()

                let ordered =
                    results
                    |> Seq.map (fun x -> x.ToContract())
                    |> Seq.sortByDescending (fun x -> x.Start)
                    |> Seq.toArray

                return { FoosballGames = ordered }
            }
